api/src/main/java/de/learnlib/api/SUL.java

```diff
@@ -87,4 +88,53 @@ default boolean canFork() {
     default SUL<I, O> fork() throws UnsupportedOperationException {
         throw new UnsupportedOperationException();
     }
+
+    /**
+     * Returns whether this SUL can retrieve the state of the system.
+     *
+     * If this method returns {@code false} then {@link #getState()} must throw an
+     * {@link UnsupportedOperationException}.
+     *
+     * @return whether the state can be retrieved.
+     */
+    default boolean canRetrieveState() {
+        return false;
+    }
+
+    /**
+     * Returns the current state of the system.
+     *
+     * Implementation note: it is important that the returned Object has a well-defined {@link Object#equals(Object)}
+     * method, and a good {@link Object#hashCode()} function.
+     *
+     * If this methods throws {@link UnsupportedOperationException}s then {@link #canRetrieveState()} must return
+     * {@code false}.
+     *
+     * @return the current state of the system.
+     *
+     * @throws UnsupportedOperationException if the current state can not be retrieved.
+     */
+    @Nonnull
+    default Object getState() throws UnsupportedOperationException {
+        throw new UnsupportedOperationException();
+    }
```

Comment:
Well `Object` exposes exactly what is necessary, the `hashcode()` and `equals()` method. The issue is that a unique identifier can not always be guaranteed (a hash collision), hence the need for `equals()`.

Reply:
To me , you still want a unique state identifier and merely compensate for the fact that `hashcode` does not guarantee this by falling back to `equals`. I think, the interface should specifiy what is actually required and it doesn't seem, you care about the actual system state other than for comparison (correct me, if I'm wrong). Wouldn't returning a primitive type (e.g. `int` or `long`) also allow you to clean up the `deepCopies()` method, because you wouldn't have to deal with actual state references that could be altered by `pre()`/`step()`/`post()` calls?

The only downside I can see is, that we delegate this uniqueness constraint to user-land code (which may be more cumbersome to implement). But I think, this can be solved in a rather generic way,  by using a state cache (e.g. in an abstract class that people can extend)

```java
final Map<Object, Integer> stateIds = new HashMap<>();

public int getStateIdentifier() {
    return stateIds.computeIfAbsent(getInternalState(), k -> stateIds.size())
}
```

Here, the HashMap implementation already deals with hashcode collisions and actual object identity (so you wouldn't have to change your current implementations returning an `Object`)

Just a thoght, interested in your opinion.

Reply:
The interface does specify exactly what is required, no? Namely a good `hashcode()` function and a  proper `equals()` function, or do you mean to what end?
I agree that having something other than `Object` could be more descriptive, but still we would use `equals()` or the operator `==` to test for state equivalence. And you are right I don't care what the state actually is.
I can tell your where the use case is coming from where `deepCopies() == false`. I use it for the RERS challenge, see the ProblemSUL interface: https://github.com/Meijuh/RERS2017/blob/master/src/main/java/nl/utwente/fmt/rers/ProblemSUL.java, and an example `Problem` instance: https://github.com/Meijuh/RERS2017-seq-problems/blob/master/src/main/java/nl/utwente/fmt/rers/problems/seq/Problem9.java. In these larger problem instances a state may be a few mega bytes large, and storing thousands can be a major issue? My reasoning is that it is cheaper to just replay a trace to bring the system to a particular state (hence it must be forkable). So you sacrifice time for memory. If you get rid of `SUL#deepCopies()` then the user does not have the option to sacrifice time for memory and is obligated to always build the map you suggest.

If the user specifies that `deepCopies() == true` then the user may let `getState()` return an `Integer`, or `Long`, but I don't care since in case `deepCopies() == true` I am only interested in the `equals()` method. However, we could change code so that operator `==` is used when `getState()` returns an `int`, or `long`, I suppose that would be a bit faster than calling `equals()`? However, performance is not an issue here.

An option would be to specialize the `SUL` for black-box checking in two ways:
1. in one specialization; states are always deep copies and `getState()` must return a unique `int`.
1. in the other; states are not guaranteed to be deep copies and `getState()` returns an `Object`. In this case the `SUL` could also be parameterized by say `S`, and `getState()` returns `S` instead of `Object`. `S` could have a method defined such as `compareMeToAnotherState(S s)`, but this is exactly the purpose of `Object#equals()`. So I don't really see the advantage of parameterizing the `SUL` specialization here.

I think I favor the way it is now, except for moving the added methods to a single specialization of the `SUL`. What would be a good name for this SUL? Glass-box, white-box?